--- hgwebdir_mod.py.orig	2014-01-02 07:46:39.000000000 +0400
+++ hgwebdir_mod.py	2015-04-15 17:55:03.774011262 +0300
@@ -368,6 +368,24 @@
                            archives=archivelist(u, "tip", url))
 
                 seenrepos.add(name)
+
+                # > this mercurial is forked (simno)
+                # shows current state of repository if it's not newest one
+                revinfo = lambda ctx: dict(
+                    rev = "%s:%s" % (ctx.rev(), ctx),
+                    branch = ctx.branch(),
+                    tags = ctx.tags(),
+                    bookmarks = ctx.bookmarks(),
+                    description = "%s: %s" % (ctx.user().strip(), ctx.description()))
+
+                if "default" in r:
+                    row["default"] = [dict(tags="", bookmarks="", description="")]
+                    row["default"][0].update(revinfo(r["default"]))
+                    row["tip"] = []
+                    if r["tip"] != r["default"]:
+                        row["tip"].append(revinfo(r["tip"]))
+                # end of hack
+
                 yield row
 
         sortdefault = None, False
